Check Voltage and sending to CANBUS

javierpadila2001 opened the pull request:
Adding a check voltage to the BSP. Will still need to grab alex code for the voltage value.

javierpadila2001 committed:
Adding a check voltage to the BSP

Will still need to grab alex code for the voltage value

ClarkPoon commented on Apps/Inc/main.c:
```diff
@@ -0,0 +1,12 @@
+#include "config.h"
+// #include "ALEX FOLDER " ==> will hold checkVoltage() function
+
+while(checkVolatge() > 0){
+ BSP_CAN_Write(0x10B, checkVoltage() )     //send message to CAN Controller 
+}
```
In C, you need to put your code inside of a function called `main` like this:
```C
int main(void) {
    // do stuff
    . . . 
}
```

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -103,16 +103,13 @@ void BSP_CAN_Init(void) {
-	for(int i = 0; i < length; i++){
-        TxMessage.Data[i] = data[i];
-	}
+    TxMessage.StdId = Voltage;
```
You are just overwriting the CAN ID here. Look at the definition of the `CanTxMsg` struct in `stm32f4xx_can.h` to see the different fields in this struct

ClarkPoon replied:
You are setting the `StdId` field of the struct `TxMessage` struct twice. See the code below (copy pasted from `BSP/STM32F413/STM32F4xx_StdPeriph_Driver/Inc/stm32f4xx_can.h` for information on what the different fields in the struct are.

```C
/** 
  * @brief  CAN Tx message structure definition  
  */
typedef struct
{
  uint32_t StdId;  /*!< Specifies the standard identifier.
                        This parameter can be a value between 0 to 0x7FF. */

  uint32_t ExtId;  /*!< Specifies the extended identifier.
                        This parameter can be a value between 0 to 0x1FFFFFFF. */

  uint8_t IDE;     /*!< Specifies the type of identifier for the message that 
                        will be transmitted. This parameter can be a value 
                        of @ref CAN_identifier_type */

  uint8_t RTR;     /*!< Specifies the type of frame for the message that will 
                        be transmitted. This parameter can be a value of 
                        @ref CAN_remote_transmission_request */

  uint8_t DLC;     /*!< Specifies the length of the frame that will be 
                        transmitted. This parameter can be a value between 
                        0 to 8 */

  uint8_t Data[8]; /*!< Contains the data to be transmitted. It ranges from 0 
                        to 0xFF. */
} CanTxMsg;
```

ClarkPoon commented on Drivers/CANbus.c:
```diff
@@ -0,0 +1,75 @@
+#include "CANbus.h"
```
I don't think we need this file, since you are just using `BSP_CAN_Write()` to send the CAN messages

javierpadila2001 replied:
So I would just delete the entire CANBUS.c? (checking before I make action)

javierpadila2001 reviewed (commented):
request completed!

javierpadila2001 committed:
Check Voltage and sending to CANBUS

run through my code again!

ClarkPoon commented on .vscode/settings.json:
```diff
@@ -12,7 +12,10 @@
         "temperature.h": "c",
         "math.h": "c",
-        "bsp_adc.h": "c"
+        //"bsp_adc.h": "c",
+        "basp_adc.h": "c",
```
Why did you do this? Did you manually edit this file?

javierpadila2001 replied:
deleted it line 15

ClarkPoon commented on Apps/Inc/main.c:
```diff
@@ -0,0 +1,14 @@
+// #include "ALEX FOLDER " ==> will hold checkVoltage() function
+
+int main(void){
+while(checkVolatge() > 0){
+ BSP_CAN_Write(0x10B, checkVoltage() )     //send message to CAN Controller 
+    }
```
```suggestion
    while(checkVolatge() > 0){
        BSP_CAN_Write(0x10B, checkVoltage() )     //send message to CAN Controller 
    }
```
Generally, we indent code blocks with 4 spaces like this. (This is style, not syntax)

javierpadila2001 replied:
Got it!

ClarkPoon commented on BSP/Template/Src/BSP_CAN.c:
```diff
@@ -12,16 +12,17 @@ void BSP_CAN_Init(void) {
- * @param   length : num of bytes of data to be transmitted. This must be <= 8 bytes or else the rest of the message is dropped.
- * @return  0 if module was unable to transmit the data onto the CAN bus. Any other value indicates data was transmitted.
+* @param   id : Message of ID. Also indicates the priority of message. The lower the value, the higher the priority.
+* @param   Voltage : the voltage value at that moment in a float
+* @return  0 if module was unable to transmit the data onto the CAN bus. Any other value indicates data was transmitted.
  */
-uint8_t BSP_CAN_Write(uint32_t id, uint8_t data[8], uint8_t length) {
+uint8_t BSP_CAN_Write(uint32_t id, float Voltage) {
     
     // TODO: Transmit the data onto the CAN bus with the specified ID.
-
-    return 0;
+    TxMessage.StdId = id;
+    TxMessage.StdId = Voltage;
+	
+	return CAN_Transmit(CAN1, &TxMessage);
 }
```
This is supposed to be the template file

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -1,144 +0,0 @@
-        // TODO: do stuff
-        RxFlag = true;
-    }
-}
```
Don't delete this file. You should edit this file to make any changes to the CAN BSP

ClarkPoon commented on .vscode/settings.json:
```diff
@@ -12,7 +12,10 @@
         "temperature.h": "c",
         "math.h": "c",
-        "bsp_adc.h": "c"
+        //"bsp_adc.h": "c",
+        "basp_adc.h": "c",
```
```suggestion
        "bsp_adc.h": "c",
```

javierpadila2001 committed:
Final voltage check code

grabbing voltage and deciding what CAN message should be sent

javierpadila2001 reviewed (commented):
Grabbing voltage and sending data to CAN

ClarkPoon commented on .vscode/settings.json:
```diff
@@ -12,7 +12,9 @@
         "temperature.h": "c",
         "math.h": "c",
-        "bsp_adc.h": "c"
+        "basp_adc.h": "c",
```
```suggestion
        "bsp_adc.h": "c",
```

ClarkPoon commented on Drivers/CANbus.c:
```diff
@@ -0,0 +1,75 @@
+
```
You can delete this file (but do not delete `BSP_CAN.c`)

javierpadila2001 committed:
Voltage Request

This is the code that look at the voltage and shares that information with the CANBUS. CANBUS will make action for f Voltage is above/below

manthanand commented on Apps/Inc/main.c:
```diff
@@ -0,0 +1,14 @@
+
+int main(void){
+    while(checkVolatge() > 0){
+        BSP_CAN_Write(0x10B, checkVoltage() )     //send message to CAN Controller 
```
There should be a semicolon here. Also this file should be in Apps/Src so replace the main.c file that is in Src and place this in there instead.

manthanand replied:
Also, the data ID for voltage data is 0x104

manthanand commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -102,17 +102,18 @@ void BSP_CAN_Init(void) {
-	for(int i = 0; i < length; i++){
-        TxMessage.Data[i] = data[i];
-	}
+    floatTo4Bytes(Voltage, TxMessage.Data[4]);
```
floatTo4Bytes is a function that is only declared in CANBus.c because it is static meaning it is local to that file. You would have to copy that function over to this file since we are not using CANBus.c and CANBus.h anymore.

javierpadila2001 replied:
I think I did this right? if not can you explain a little better. I'm on BSP_CAN.c I added those lines. but made the green line a comment since it is static I cannot just call on it like that.

javierpadila2001 committed:
mainloop_BSP

sending request to see if everything is working

manthanand commented:
Since the main.c file is an Src file, move it from Apps/Inc to Apps/Src. That way our code can stay organized and it won't confuse anybody.

ClarkPoon commented on Apps/Inc/main.c:
```diff
@@ -0,0 +1,14 @@
+// #include "ALEX FOLDER " ==> will hold checkVoltage() function
+
+int main(void){
+    while(checkVolatge() > 0){
```
Why is there a conditional here? We don't want to ever break out of this loop and reach the end of `main()` or the microcontroller will crash.

javierpadila2001 replied:
Thought we wanted it check if the voltage was in the wrong direction, thus immediately alert the driver. changed it to always be true

ClarkPoon commented on Apps/Inc/main.c:
```diff
@@ -0,0 +1,14 @@
+
+int main(void){
+    while(checkVolatge() > 0){
+        BSP_CAN_Write(0x10B, checkVoltage());     //send message to CAN Controller 
```
Can you change `0x10b` to a `#define`? I think it will make the code more readable

javierpadila2001 replied:
so I change the top to say #define "0x1b" ? then leave the 0x10b there ?

manthanand replied:
#define ALLCAPSVARIABLE 0x1b

ClarkPoon commented on Apps/Src/main.c:
```diff
@@ -4,9 +4,16 @@
  */
 
-#include "common.h"
+/* Copyright (c) 2020 UT Longhorn Racing Solar */
```
See my comments on `Apps/Inc/main.c`. Note that you will want to use this main.c and delete the other one

javierpadila2001 replied:
use Apps/Inc/main.c and delete Apps/ Inc/ main.c

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -102,17 +102,21 @@ void BSP_CAN_Init(void) {
-uint8_t BSP_CAN_Write(uint32_t id, uint8_t data[8], uint8_t length) {
+uint8_t BSP_CAN_Write(uint32_t id, float Voltage) {
+    
+    // TODO: Transmit the data onto the CAN bus with the specified ID.
```
It looks like this has already been implemented. If so, you can remove the TODO statement

javierpadila2001 replied:
comment deleted

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -140,8 +144,9 @@ void CAN1_RX0_IRQHandler(void)
         RxFlag = true;
     }
-}
+}
```
Can you add a blank line at the end of the file? The compiler will get mad and give a warning if you don't

javierpadila2001 replied:
got it!

javierpadila2001 committed:
SB for BPS

Should be completed

javierpadila2001 committed:
build runs

BSP

manthanand commented on BSP/STM32F413/Src/main.c:
```diff
@@ -0,0 +1,16 @@
+
+/* Copyright (c) 2020 UT Longhorn Racing Solar */
+
+#define BSPVARIABLE 0x10b
```
Probably calL this something more descriptive like VOLTAGECANID

ClarkPoon replied:
Yes, definitely. This is a bad name for a `#define`, since it tells you nothing about what it is used for (and it is not even a variable).

manthanand commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -102,17 +102,20 @@ void BSP_CAN_Init(void) {
-    TxMessage.DLC = length;
-	for(int i = 0; i < length; i++){
-        TxMessage.Data[i] = data[i];
+    for(int i = 0; i < Voltage; i++){
```
This for loop doesn't look right. You are looping through the number of times of the voltage value, not the length of the actual data.

manthanand commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -102,17 +102,20 @@ void BSP_CAN_Init(void) {
+      // Voltage[i] = TxMessage.Voltage[i];
 	}
+    // floatTo4Bytes(Voltage, TxMessage.Data[4]);
+    Voltage = TxMessage.Data[4];
```
Over here, you are changing the value of voltage. Is that what you meant to do or did you mean to change the value of TxMessage.Data.

manthanand commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -102,17 +102,20 @@ void BSP_CAN_Init(void) {
+    for(int i = 0; i < Voltage; i++){
+      // Voltage[i] = TxMessage.Voltage[i];
 	}
+    // floatTo4Bytes(Voltage, TxMessage.Data[4]);
```
You should use this function because CAN_Transmit can only take as input 4 bytes.

javierpadila2001 committed:
Retrieving the PA0 value

getting the voltage value, storing it into  check voltage, it will then jump to main where it uses the check Voltage to send to CANBUS

manthanand committed:
Created ADC functions. Need to fix CAN Write function. Need to check how ADC data is received to do proper conversion of voltage.

manthanand committed:
Implemented ADC conversion and CAN write functions

ClarkPoon commented on BSP/STM32F413/Src/main.c:
```diff
@@ -0,0 +1,18 @@
+ /* Program for UTSVT BeVolt's Battery Protection System*/
```
Why is `main.c` in the BSP folder?

ClarkPoon commented on Drivers/Inc/CANbus.h:
```diff
@@ -3,7 +3,7 @@
 #define __CANBUS_H
 
-#include "common.h"
+//#include "common.h"
```
If this header isn't needed, we should just delete it

ClarkPoon commented on Apps/Src/main.c:
```diff
@@ -1,12 +0,0 @@
-/* Copyright (c) 2020 UT Longhorn Racing Solar */
```
Why would you delete this file?

ClarkPoon commented on BSP/Inc/BSP_ADC.h:
```diff
@@ -0,0 +1,10 @@
+#include "config.h"
+#include "stm32f4xx.h"
+
+//Initialize ADC
```
Can we comment the functions with the same comment style as in `BSP/Inc/BSP_CAN.h`?

ClarkPoon replied:
Can you fix up the comments in this file @tasniabari ?

ClarkPoon commented on BSP/STM32F413/Src/BSP_ADC.c:
```diff
@@ -0,0 +1,50 @@
+ /* Program for UTSVT BeVolt's Battery Protection System*/
```
Delete this comment (it is incorrect)

ClarkPoon commented on BSP/STM32F413/Src/BSP_ADC.c:
```diff
@@ -0,0 +1,50 @@
+	ADC_SoftwareStartConv(ADC1);
+}
+
+float BSP_ADC_UpdateMeasurements(void){
```
This function should have a comment

ClarkPoon commented on BSP/STM32F413/Src/BSP_ADC.c:
```diff
@@ -0,0 +1,50 @@
+ /* Copyright (c) 2021 UT Longhorn Racing Solar */
+#include "BSP_ADC.h"
+
+void BSP_ADC_Init(){
```
This function should have a comment

ClarkPoon replied:
@tasniabari after you add a comment for this function to BSP_ADC.h, can you copy-paste the function comment here?

ClarkPoon commented on BSP/STM32F413/Src/BSP_ADC.c:
```diff
@@ -0,0 +1,50 @@
+
+float BSP_ADC_UpdateMeasurements(void){
+    uint16_t val = ADC_GetConversionValue(ADC1); //0x000 means 14V and 0xFFF means 0V (12-bit ADC)
+    float mV = (val * 3300) >> 12;
```
Why are we using bit shifts to do division, if the result is a `float`?

ClarkPoon commented on BSP/STM32F413/Src/BSP_ADC.c:
```diff
@@ -0,0 +1,50 @@
+float BSP_ADC_UpdateMeasurements(void){
+    uint16_t val = ADC_GetConversionValue(ADC1); //0x000 means 14V and 0xFFF means 0V (12-bit ADC)
+    float mV = (val * 3300) >> 12;
+    return 1400 - mV; //return value in mV
```
```suggestion
    return 1400.0F - mV; //return value in mV
```

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -10,6 +10,8 @@ static CanRxMsg RxMessage;
 
 static bool RxFlag = false;
 
+static void floatTo4Bytes(uint8_t val, uint8_t bytes_array[4]);
```
Why doesn't this take a float as a parameter???

ClarkPoon replied:
Can we delete this function, since we aren't using it?

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -102,17 +104,15 @@ void BSP_CAN_Init(void) {
+* @return  0 if module was unable to transmit the data onto the CAN bus. Any other value indicates data was transmitted.
  */
-uint8_t BSP_CAN_Write(uint32_t id, uint8_t data[8], uint8_t length) {
+uint8_t BSP_CAN_Write(uint32_t id, float Voltage) {
```
It is probably a good idea to preserve the old `BSP_CAN_Write` in case we ever want to send any other types of CAN messages from this board

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -102,17 +104,15 @@ void BSP_CAN_Init(void) {
-	for(int i = 0; i < length; i++){
-        TxMessage.Data[i] = data[i];
-	}
+    floatTo4Bytes(Voltage, &TxMessage.Data[4]);
+
 	return CAN_Transmit(CAN1, &TxMessage);
```
```suggestion
    memcpy(TxMessage.Data, &Voltage, sizeof(Voltage));
    return CAN_Transmit(CAN1, &TxMessage);
```

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -140,8 +140,29 @@ void CAN1_RX0_IRQHandler(void)
+
+}
+
+static void floatTo4Bytes(uint8_t val, uint8_t bytes_array[4]) {
```
I'm not convinced we need this function

ClarkPoon replied:
I think we can delete this function if we aren't using it

ClarkPoon commented on BSP/STM32F413/Src/main.c:
```diff
@@ -0,0 +1,18 @@
+    while(1) {
+        err = BSP_CAN_Write(BSPVARIABLE, BSP_ADC_UpdateMeasurements()); //send message to CAN Controller 
+        if (err == 0){/*ERROR HANDLING HERE*/}
+    }
```
We will want a delay here, or we will spam the CAN bus and stop the other systems from using it

ClarkPoon committed:
added cast to float

kevingonz19 committed:
altered BSP_ADC.c to user int32 values while also using mV instead of V to avoid precision loss. Clark then altered files which used the old Update_measurements function from float to int variable types. Finally, started two test files for ADC and CAN

tasniabari committed:
Added CAN write and read to main

tasniabari committed:
Modified CAN read function, added some checks for CAN write to make sure message is transmitted

kevingonz19 committed:
The important math to calculate the suplemental battery reading is in Test_ADC_Val. That approximates the supp battery output pretty well, although not perfectly. It will be included in the main function once the main function is written.

kevingonz19 committed:
Merge remote-tracking branch 'refs/remotes/origin/mainloop_BSP' into mainloop_BSP

tasniabari committed:
Fixed Transmission and Receiving of Messages on loopback mode. Changed frame transmission number to 2 to read voltage data.

tasniabari committed:
changes to makefile

ClarkPoon commented on Apps/Src/main.c:
```diff
@@ -1,12 +1,39 @@
+int main(void){
+    BSP_ADC_Init();
+    BSP_CAN_Init();
+    volatile uint8_t mailBoxNum;
```
There is a lot of debugging code in here. I think it is fine to leave it in for now, since we don't know if this will work when we try to integrate with the display board. After we know this works, we should remove all fo the debugging code to clean it up

tasniabari replied:
I rearranged some lines in the variable declarations and while loop , so now, what needs to be removed after integration is clearly marked.

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -80,15 +83,21 @@ void BSP_CAN_Init(void) {
-    TxMessage.DLC = 1;
+    TxMessage.DLC = 2;         //This needs to be 2, because we need 2 frames because the battery voltage data is 4 bytes, and each frame is 2 bytes
+
+    //idk if this will change anything
+    //for(int i = 0; i < 8; i++){TxMessage.Data[i] = 0;}
+	//return CAN_Transmit(CAN1, &TxMessage);
```
Can we delete this?

ClarkPoon commented on BSP/STM32F413/Src/BSP_ADC.c:
```diff
@@ -0,0 +1,50 @@
+
+	ADC_SoftwareStartConv(ADC1);
+}
+// Recieve ADC value as val, and convert to a value between 0 and 14,000 mV to avoid using floats
```
@tasniabari can you also change this function to the style used in BSP_CAN? This should also match the function comment in BSP_ADC.h

tasniabari replied:
Done, I also got rid of a lot of the comments within the ADC functions and I'd like to add them back at some point.

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -47,7 +50,7 @@ void BSP_CAN_Init(void) {
     /* CAN cell init */
     CAN_InitStructure.CAN_TTCM = DISABLE;
-    CAN_InitStructure.CAN_ABOM = DISABLE;
+    CAN_InitStructure.CAN_ABOM = DISABLE; //was DISABLE
```
can we remove this comment?

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -3,19 +3,22 @@
  * @return  None
  */
-void BSP_CAN_Init(void) {
+void BSP_CAN_Init(void) { //orignally was void return type
```
can we remove this comment?

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -80,15 +83,21 @@ void BSP_CAN_Init(void) {
 
     /* Receive Structure preparation */
-    RxMessage.StdId = 0x00;
+    RxMessage.StdId = 0x00; //was 0
```
Can we delete this comment?

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -98,21 +107,32 @@ void BSP_CAN_Init(void) {
     NVIC_InitStructure.NVIC_IRQChannelSubPriority = 0x0;
     NVIC_InitStructure.NVIC_IRQChannelCmd = ENABLE;
     NVIC_Init(&NVIC_InitStructure);	
+
+
+    //return CANinitSuccess;
+    
```
Can we delete this?

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -80,15 +83,21 @@ void BSP_CAN_Init(void) {
     RxMessage.DLC = 0;
     RxMessage.FMI = 0;
 
+    
+
```
Can we delete this?

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -98,21 +107,32 @@ void BSP_CAN_Init(void) {
- * @param   length : num of bytes of data to be transmitted. This must be <= 8 bytes or else the rest of the message is dropped.
- * @return  0 if module was unable to transmit the data onto the CAN bus. Any other value indicates data was transmitted.
+* @param   id : Message of ID. Also indicates the priority of message. The lower the value, the higher the priority.
+* @param   Voltage : the voltage value at that moment in a float
```
This comment needs to be updated, since we are sending the voltage as an integer with units of millivolts

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -98,21 +107,32 @@ void BSP_CAN_Init(void) {
-	for(int i = 0; i < length; i++){
-        TxMessage.Data[i] = data[i];
-	}
+       //floatTo4Bytes(Voltage, &TxMessage.Data[4]);
```
Can we remove the commented out code in this function?

tasniabari replied:
Deleted that and some other commented out stuff in the BPS_CAN_Write function.

ClarkPoon commented on BSP/STM32F413/Src/BSP_CAN.c:
```diff
@@ -124,24 +144,49 @@ uint8_t BSP_CAN_Write(uint32_t id, uint8_t data[8], uint8_t length) {
+uint8_t BSP_CAN_Read(uint32_t *id, uint8_t *data) { //originally returns uint8_t
+
+    
+    //if(RxFlag){
```
what was `RxFlag` being used for, and is it safe to remove it?

tasniabari replied:
I believe this flag was to check if the message being sent was meaningful or not. I reincluded this and modified the hardware interrupt so it would set this flag to "True" whenever it's called, since the message being sent is probably not junk if this interrupt is even invoked.

tasniabari committed:
Added useful info for integration, as well as marked code to be deleted once testing and integration is complete. Cleaned up comments in ADC files. This code has been tested with logic analyzer and it returns inputs as expected.
